File: swim-system-java/swim-core-java/swim.codec/src/main/java/swim/codec/Mark.java
// Copyright 2015-2019 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.codec;

import java.util.Objects;
import swim.util.Murmur3;

/**
 * Description of a source position, identified by byte offset, line, and
 * column number, with an optional note.
 */
public final class Mark extends Tag implements Comparable<Mark> {
  final long offset;
  final int line;
  final int column;
  final String note;

  Mark(long offset, int line, int column, String note) {
    this.offset = offset;
    this.line = line;
    this.column = column;
    this.note = note;
  }

  /**
   * Returns the zero-based byte offset of this position.
   */
  public long offset() {
    return this.offset;
  }

  /**
   * Returns the one-based line number of this position.
   */
  public int line() {
    return this.line;
  }

  /**
   * Returns the one-based column number of this position.
   */
  public int column() {
    return this.column;
  }

  /**
   * Returns the note attached to the marked position, or {@code null} if this
   * position has no attached note.
   */
  public String note() {
    return this.note;
  }

  /**
   * Returns {@code this} position, if its byte offset is less than or equal
   * to {@code that} position; otherwise returns {@code that} position.
   */
  public Mark min(Mark that) {
    if (this.offset <= that.offset) {
      return this;
    } else {
      return that;
    }
  }

  /**
   * Returns {@code this} position, if its byte offset is greater than or equal
   * to {@code that} position; otherwise returns {@code that} position.
   */
  public Mark max(Mark that) {
    if (this.offset >= that.offset) {
      return this;
    } else {
      return that;
    }
  }

  @Override
  public Mark start() {
    return this;
  }

  @Override
  public Mark end() {
    return this;
  }

  @Override
  public Tag union(Tag other) {
    if (other instanceof Mark) {
      final Mark that = (Mark) other;
      if (this.offset == that.offset && this.line == that.line
          && this.column == that.column) {
        return this;
      } else {
        return Span.from(this, that);
      }
    } else if (other instanceof Span) {
      final Span that = (Span) other;
      final Mark start = min(that.start);
      final Mark end = max(that.end);
      if (start == that.start && end == that.end) {
        return that;
      } else {
        return Span.from(start, end);
      }
    }
    throw new UnsupportedOperationException(other.toString());
  }

  @Override
  public Mark shift(Mark mark) {
    final long offset = this.offset + (this.offset - mark.offset);
    final int line = this.line + (this.line - mark.line);
    int column = this.column;
    if (line == 1) {
      column += (this.column - mark.column);
    }
    if (offset == this.offset && line == this.line && column == this.column) {
      return this;
    } else {
      return Mark.at(offset, line, column, this.note);
    }
  }

  @Override
  public int compareTo(Mark that) {
    return Long.compare(this.offset, that.offset);
  }

  @Override
  public boolean equals(Object other) {
    if (this == other) {
      return true;
    } else if (other instanceof Mark) {
      final Mark that = (Mark) other;
      return this.offset == that.offset && this.line == that.line
          && this.column == that.column && Objects.equals(this.note, that.note);
    }
    return false;
  }

  @Override
  public int hashCode() {
    if (hashSeed == 0) {
      hashSeed = Murmur3.seed(Mark.class);
    }
    return Murmur3.mash(Murmur3.mix(Murmur3.mix(Murmur3.mix(Murmur3.mix(hashSeed,
        Murmur3.hash(this.offset)), this.line), this.column), Murmur3.hash(this.note)));
  }

  @Override
  public void display(Output<?> output) {
    Format.displayInt(this.line, output);
    output = output.write(':');
    Format.displayInt(this.column, output);
    if (this.note != null) {
      output = output.write(": ").write(this.note);
    }
  }

  @Override
  public void debug(Output<?> output) {
    output = output.write("Mark").write('.').write("at").write('(');
    Format.debugLong(this.offset, output);
    output = output.write(", ");
    Format.debugInt(this.line, output);
    output = output.write(", ");
    Format.debugInt(this.column, output);
    if (this.note != null) {
      output = output.write(", ");
      Format.debugString(this.note, output);
    }
    output = output.write(')');
  }

  @Override
  public String toString() {
    return Format.display(this);
  }

  private static int hashSeed;
  private static Mark zero;

  /**
   * Returns a {@code Mark} at byte offset {@code 0}, line {@code 1}, and
   * column {@code 1}, with no attached note.
   */
  public static Mark zero() {
    if (zero == null) {
      zero = new Mark(0L, 1, 1, null);
    }
    return zero;
  }

  /**
   * Returns a new {@code Mark} at the given zero-based byte {@code offset},
   * one-based {@code line} number, and one-based {@code column} number,
   * with the attached {@code note}.
   */
  public static Mark at(long offset, int line, int column, String note) {
    return new Mark(offset, line, column, note);
  }

  /**
   * Returns a new {@code Mark} at the given zero-based byte {@code offset},
   * one-based {@code line} number, and one-based {@code column} number,
   * with no attached note.
   */
  public static Mark at(long offset, int line, int column) {
    return at(offset, line, column, null);
  }
}

File: swim-system-js/swim-core-js/@swim/codec/main/Mark.ts
// Copyright 2015-2020 SWIM.AI inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {Murmur3} from "@swim/util";
import {Tag} from "./Tag";
import {Output} from "./Output";

/**
 * Description of a source position, identified by byte offset, line, and
 * column number, with an optional note.
 */
export class Mark extends Tag {
  /** @hidden */
  readonly _offset: number;
  /** @hidden */
  readonly _line: number;
  /** @hidden */
  readonly _column: number;
  /** @hidden */
  readonly _note: string | null;

  /** @hidden */
  constructor(offset: number, line: number, column: number, note: string | null) {
    super();
    this._offset = offset;
    this._line = line;
    this._column = column;
    this._note = note;
  }

  /**
   * Returns the zero-based byte offset of this position.
   */
  offset(): number {
    return this._offset;
  }

  /**
   * Returns the one-based line number of this position.
   */
  line(): number {
    return this._line;
  }

  /**
   * Returns the one-based column number of this position.
   */
  column(): number {
    return this._column;
  }

  /**
   * Returns the note attached to the marked position, or `null` if this
   * position has no attached note.
   */
  note(): string | null {
    return this._note;
  }

  /**
   * Returns `this` position, if its byte offset is less than or equal to
   * `that` position; otherwise returns `that` position.
   */
  min(that: Mark): Mark {
    if (this._offset <= that._offset) {
      return this;
    } else {
      return that;
    }
  }

  /**
   * Returns `this` position, if its byte offset is greater than or equal to
   * `that` position; otherwise returns `that` position.
   */
  max(that: Mark): Mark {
    if (this._offset >= that._offset) {
      return this;
    } else {
      return that;
    }
  }

  start(): Mark {
    return this;
  }

  end(): Mark {
    return this;
  }

  union(that: Tag): Tag {
    if (that instanceof Mark) {
      if (this._offset === that._offset && this._line === that._line
          && this._column === that._column) {
        return this;
      } else {
        return Tag.Span.from(this, that);
      }
    } else if (that instanceof Tag.Span) {
      const start = this.min(that._start);
      const end = this.max(that._end);
      if (start === that._start && end === that._end) {
        return that;
      } else {
        return Tag.Span.from(start, end);
      }
    }
    throw new Error(that.toString());
  }

  shift(mark: Mark): Mark {
    const offset = this._offset + (this._offset - mark._offset);
    const line = this._line + (this._line - mark._line);
    let column = this._column;
    if (line === 1) {
      column += (this._column - mark._column);
    }
    if (offset === this._offset && line === this._line && column === this._column) {
      return this;
    } else {
      return Mark.at(offset, line, column, this._note);
    }
  }

  equals(that: unknown): boolean {
    if (this === that) {
      return true;
    } else if (that instanceof Mark) {
      return this._offset === that._offset && this._line === that._line
          && this._column === that._column && this._note === that._note;
    }
    return false;
  }

  hashCode(): number {
    if (Mark._hashSeed === void 0) {
      Mark._hashSeed = Murmur3.seed(Mark);
    }
    return Murmur3.mash(Murmur3.mix(Murmur3.mix(Murmur3.mix(Murmur3.mix(Mark._hashSeed,
        this._offset), this._line), this._column), Murmur3.hash(this._note)));
  }

  display(output: Output): void {
    Tag.Format.displayNumber(this._line, output);
    output = output.write(58/*':'*/);
    Tag.Format.displayNumber(this._column, output);
    if (this._note !== null) {
      output = output.write(58/*':'*/).write(32/*' '*/).write(this._note);
    }
  }

  debug(output: Output): void {
    output = output.write("Mark").write(".").write("at").write("(");
    Tag.Format.debugNumber(this._offset, output);
    output = output.write(", ");
    Tag.Format.debugNumber(this._line, output);
    output = output.write(", ");
    Tag.Format.debugNumber(this._column, output);
    if (this._note !== null) {
      output = output.write(", ");
      Tag.Format.debugString(this._note, output);
    }
    output = output.write(")");
  }

  toString(): string {
    return Tag.Format.display(this);
  }

  private static _hashSeed?: number;
  private static _zero?: Mark;

  /**
   * Returns a `Mark` at byte offset `0`, line `1`, and column `1`, with no
   * attached note.
   */
  static zero(): Mark {
    if (Mark._zero === void 0) {
      Mark._zero = new Mark(0, 1, 1, null);
    }
    return Mark._zero;
  }

  /**
   * Returns a new `Mark` at the given zero-based byte `offset`, one-based
   * `line` number, and one-based `column` number, with an optionally attached
   * `note`.
   */
  static at(offset: number, line: number, column: number, note: string | null = null): Mark {
    return new Mark(offset, line, column, note);
  }
}
Tag.Mark = Mark;
